src/components/Store.react.js

```diff
@@ -125,11 +133,11 @@ export default class Store extends React.Component {
     constructor(props) {
         super(props);
 
-        if (props.storage_type === 'local') {
-            this._backstore = _localStore;
+        if (props.storage_type === 'local' && localStorageTest()) {
+            this._backstore = new WebStore(window.localStorage);
         } else if (props.storage_type === 'session') {
-            this._backstore = _sessionStore;
-        } else if (props.storage_type === 'memory') {
+            this._backstore = new WebStore(window.sessionStorage);
+        } else {
```

Comment:
This change is unlike the similarly themed table change (https://github.com/plotly/dash-table/pull/677) that impact logging but no user-facing feature.

This component defaults to `memory` storage and `local` needs to be explicitly set to potentially run into issues with iframes. This will silently break expectations and will look like a bug to most users as the expectation will be for `local` to store information across sessions and `session` to store information across page reloads (e.g. docs example in https://dash.plot.ly/dash-core-components/store).

I think the correct approach here is to **not** try and be smart and **not** degrade functionality. Instead, I would suggest that we throw an error that will appear in the browser's devtools and that will be picked up by the Dash devtools with `debug=True`. The application developers will then be able to fix their code or its environment settings to either respect the sandbox or unrestrict localStorage as needed.

Tagging @alexcjohnson for a second opinion.

If we do decide to go the degradation way, the localStorage should at least degrade to sessionStorage instead of memory storage.

Reply:
yeah that makes sense. As long as the error happens on trying to create the component, rather than on initial script execution as it does prior to this PR.

> localStorage should at least degrade to sessionStorage instead of memory

Does `sessionStorage` still work when `localStorage` is forbidden?

Reply:
Were we validating the `storage_type` value anywhere previously?

I guess my overall desire here is to ensure we're not always attempting to use window.localStorage/window.sessionStorage if those aren't available, which we were previously doing.

How do we throw the error for dash devtools?

Reply:
I think `sessionStorage` is allowed but might be sandboxed from the parent page's `sessionStorage` based on attributes. Checking.

Reply:
I've reverted and added a fallback to sessionstorage, but I can also remove that fallback if it's okay to just error directly.

Reply:
There seems to be no way to both successfully (a) create the Store component and (b) throw an Error that will be caught by the React Error Boundary. Given that limitation, the correct implementation for now IMO would be to throw the error / fail the Store component rendering.

I will create a followup issue to look into making it possible to both (a) degrade / render a component behavior, and (b) expose the error in Dash devtools. (e.g. have a communication mechanism that does not use the Error Boundary to expose the error)

Reply:
yeah, it's OK to just throw an error during the component lifecycle - that'll make it to the devtools.

> I guess my overall desire here is to ensure we're not always attempting to use window.localStorage/window.sessionStorage if those aren't available, which we were previously doing.

That's what I was getting at with "As long as the error happens on trying to create the component, rather than on initial script execution as it does prior to this PR." Don't access any storage systems on initial execution; when you make a component, try to access the requested storage system and only that one, and allow it to throw an error if that fails.

Reply:
I removed the fallback code and localstorage test.